src/material/datepicker/datepicker-input.ts

```diff
@@ -94,21 +100,28 @@ export class MatDatepickerInputEvent<D> {
 export class MatDatepickerInput<D> implements ControlValueAccessor, OnDestroy, Validator {
   /** The datepicker that this input is associated with. */
   @Input()
-  set matDatepicker(value: MatDatepicker<D>) {
-    if (!value) {
+  set matDatepicker(datepicker: MatDatepicker<D>) {
+    if (!datepicker) {
       return;
     }
 
-    this._datepicker = value;
-    this._datepicker._registerInput(this);
-    this._datepickerSubscription.unsubscribe();
+    this._datepicker = datepicker;
+    this._model = this._datepicker._registerInput(this);
+    this._valueChangesSubscription.unsubscribe();
 
-    this._datepickerSubscription = this._datepicker._selectedChanged.subscribe((selected: D) => {
-      this.value = selected;
-      this._cvaOnChange(selected);
-      this._onTouched();
-      this.dateInput.emit(new MatDatepickerInputEvent(this, this._elementRef.nativeElement));
-      this.dateChange.emit(new MatDatepickerInputEvent(this, this._elementRef.nativeElement));
+    if (this._pendingValue) {
+      this._assignValue(this._pendingValue);
+      this._pendingValue = null;
```

Comment:
should we just move this line into `_assignValue` on the branch where we find a selection model?